pkg/controller/mutators/core/controller_core.go

```diff
@@ -0,0 +1,314 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Adder struct {
+	MutationSystem *mutation.System
+	Tracker        *readiness.Tracker
+	GetPod         func() (*corev1.Pod, error)
+	Kind           string
+	NewObj         func() client.Object
+	NewMutator     func(client.Object) (types.Mutator, error)
+}
+
+// Add creates a new Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr, a.MutationSystem, a.Tracker, a.GetPod, a.Kind, a.NewObj, a.NewMutator)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(o *opa.Client) {}
+
+func (a *Adder) InjectWatchManager(w *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(t *readiness.Tracker) {
+	a.Tracker = t
+}
+
+func (a *Adder) InjectGetPod(getPod func() (*corev1.Pod, error)) {
+	a.GetPod = getPod
+}
+
+func (a *Adder) InjectMutationSystem(mutationSystem *mutation.System) {
+	a.MutationSystem = mutationSystem
+}
+
+// newReconciler returns a new reconcile.Reconciler.
+func newReconciler(
+	mgr manager.Manager,
+	mutationSystem *mutation.System,
+	tracker *readiness.Tracker,
+	getPod func() (*corev1.Pod, error),
+	kind string,
+	newObj func() client.Object,
+	newMutator func(client.Object) (types.Mutator, error),
+) *Reconciler {
+	r := &Reconciler{
+		system:     mutationSystem,
+		Client:     mgr.GetClient(),
+		tracker:    tracker,
+		getPod:     getPod,
+		scheme:     mgr.GetScheme(),
+		reporter:   ctrlmutators.NewStatsReporter(),
+		cache:      ctrlmutators.NewMutationCache(),
+		kind:       kind,
+		newObj:     newObj,
+		newMutator: newMutator,
+		log:        logf.Log.WithName("controller").WithValues(logging.Process, fmt.Sprintf("%s_controller", strings.ToLower(kind))),
+	}
+	if getPod == nil {
+		r.getPod = r.defaultGetPod
+	}
+	return r
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler.
+func add(mgr manager.Manager, r *Reconciler) error {
+	if !*mutation.MutationEnabled {
+		return nil
+	}
+
+	// Create a new controller
+	c, err := controller.New(fmt.Sprintf("%s-controller", strings.ToLower(r.kind)), mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the mutator
+	if err = c.Watch(
+		&source.Kind{Type: r.newObj()},
+		&handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	// Watch for changes to MutatorPodStatus
+	err = c.Watch(
+		&source.Kind{Type: &statusv1beta1.MutatorPodStatus{}},
+		handler.EnqueueRequestsFromMapFunc(mutatorstatus.PodStatusToMutatorMapper(true, r.kind, util.EventPackerMapFunc())),
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Reconciler reconciles mutator objects.
+type Reconciler struct {
```

Comment:
Might be nice for `Reconciler` type declaration to be above `newReconciler()` in the code, for readability

Reply:
Depends on what you're looking for. If you want to know "how do I add a reconciler", it's better to have the `newReconciler` at the top, as consumers of this generally don't care how the reconciler is implemented.

The current order of things sticks as closely to the kubebuilder generated output as we can, which hopefully makes the code more readable to those with that context.